doc/keynames/README.md

```diff
@@ -0,0 +1,473 @@
+
+Unescaping works, by simply removing the backslashes `\` that are used as escapes.
+This applies both to `\/` and `\\` anywhere in Key Name Parts, as well as to the escape sequences that are only used at the start of Key Name Parts, e.g. `\#`.
+
+## 4. Valid and Invalid Key Names
+
+Not all of the Key Names described by the above sections are valid under all circumstances.
+You might also say, that a Key Name as defined above does not necessarily refer to a Key in the KDB.
+So while it might be a Key Name, there is no Key that actually uses this name.
+Only _Valid Key Names_ refer to a Key in the KDB and may be used by a Key as its name.
+The remaining Key Names are referred to as _Invalid Key Names_.
+
+For Unescaped Key Names, it is pretty simple:
+Unescaped Key Names are Valid Key Names, if all of the following are true:
+
+- The first byte is a valid Namespace Byte, i.e. `0x01` - `0x08`.
+- The second byte is a zero-byte `0x00`.
+- The last byte is a zero-byte `0x00`.
+
+For this reason Unescaped Names for Root Keys are 3 bytes long, with the last two bytes being zero-byte.
+Between the second and the last byte, we find the Key Name Parts separated by a zero-byte `0x00`.
+
+For Escaped Key Names it is easier to define, what makes an Invalid Key Name, than what makes a Valid Key Name, so we will go this route.
+An Escaped Key Name is considered an Invalid Key Name, if any of the following are true:
+
+- It is an empty string.
+- The last character before the invisible null terminator is a backslash `\`.
+  This is a dangling escape as we expect another character after the escape character `\`.
+- It contains a Namespace (i.e. the Namespace is not "Cascading"), but the Namespace Separator `:` is not followed by a `/`.
+  (This mainly applies to Root Keys.)
+- It contains a Namespace Separator `:`, but the substring before the first `:` is not one of: `meta`, `spec`, `proc`, `dir`, `user`, `system` and `default`.
+- One of the Key Name Parts starts with an unescaped `@`.
+  For example `/elektra/@version` is an invalid Escaped Key Name.
+  Only `/elektra/\@version` would be valid.
+- It contains an Illegal Escape Sequence (see below).
+- It contains an Invalid Array Part (see below).
+- Going from left to right one Key Name Part at a time, at any point, the number of Key Name Parts that are `..` is bigger than the number of other Key Name Parts.
+  In Non-Canonical Key Names, the `.` Key Name Parts contribute to the number of other Key Name Parts.
+
+An Escape Sequence is illegal, if (one of):
+
+- It appears at the start of a Key Name Part and the escaped character is not one of: `/`, `\`, `.`, `#`, `%` or `@`.
+- It appears anywhere else and the escaped character is not one of: `/` or `\`.
+
+An Array Part is valid, if (one of):
+
+- The whole Key Name Part is `#` or `#0`.
+- The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^64 - 1`).
```

Comment:
Elektra uses a signed integer (`kdb_long_long_t`) for array indices. I think this is so that `-1` can be used as an error code for `elektraReadArrayNumber`. `ELEKTRA_MAX_ARRAY_SIZE` has not been updated yet, if you took that as a reference.

Reply:
If I am honest, I did not look at the type of array indices. I just assumed that arrays indices use a 64-bit unsigned type, because of the text in the parentheses (`2^64 - 1`), which is `18446744073709551615` and not `9223372036854775807` (= `2^63 - 1`), according to my Python interpreter. Anyway, if we assume a signed 64-bit number, then I think we need to update the exponent:

```suggestion
- The Key Name Part is `#` followed by 1-19 digits, the first digit is not `0`, and the digits form a number less than or equal to `9223372036854775807` (= `2^63 - 1`).
```

Reply:
You're right it should be 2^63 - 1.